src/passes/OptimizeInstructions.cpp

```diff
@@ -841,6 +757,66 @@ struct OptimizeInstructions
           }
         }
       }
+      if (auto* condition = select->condition->dynCast<Unary>()) {
+        if (condition->op == EqZInt32) {
+          // flip select to remove eqz, if we can reorder
+          EffectAnalyzer ifTrue(getPassOptions(), features, select->ifTrue);
+          EffectAnalyzer ifFalse(getPassOptions(), features, select->ifFalse);
+          if (!ifTrue.invalidates(ifFalse)) {
+            select->condition = condition->value;
+            std::swap(select->ifTrue, select->ifFalse);
+          }
+        }
+      }
+      if (auto* constTrue = select->ifTrue->dynCast<Const>()) {
+        if (auto* constFalse = select->ifFalse->dynCast<Const>()) {
+          Builder builder(*getModule());
+          bool is64bits = select->type == Type::i64;
+          auto extendIfNeeded = [&](Expression* expression) {
+            return is64bits ? builder.makeUnary(ExtendUInt32, expression)
+                            : expression;
+          };
+          if (select->type == Type::i32 || is64bits) {
+            if (constTrue->type == constFalse->type &&
+                constTrue->value.getInteger() == 1LL &&
+                constFalse->value.getInteger() == 0LL) {
+              auto* condition = select->condition;
+              if (Properties::emitsBoolean(condition)) {
+                // !x ? 1 : 0   ==>   !x
+                // x <=> y ? 1 : 0   ==>   x <=> y
+                return extendIfNeeded(condition);
+              } else {
+                // expr ? 1 : 0   ==>   expr != 0
+                return extendIfNeeded(builder.makeBinary(
+                  NeInt32, condition, builder.makeConst(Literal(0))));
+              }
+            } else if (constTrue->type == constFalse->type &&
+                       constTrue->value.getInteger() == 0LL &&
+                       constFalse->value.getInteger() == 1LL) {
+              Builder builder(*getModule());
```

Comment:
in general this area of the code looks complicated. how about this:

* Check if the arms are `0, 1`. If they are, add an EqZ and flip them.
* If they are `1, 0` (either they were originally, or we flipped them) do the opts for that case.

That seems simpler, and existing opts will get rid of the EqZ whenever possible.